### PR TITLE
Live activity support 5.0.0

### PR DESCRIPTION
**Pull request by fhboswell**
# Description
## One Line Summary
Add Live Activity support to Major release 5.0.0

## Details
Live Activities are a type of interactive push notification. They were designed by Apple to enable iOS apps to provide real-time updates to their users that are visible from the lock screen and dynamic island.

```
let activity = try Activity<OneSignalWidgetAttributes>.request(
  attributes: attributes,
  contentState: contentState,
  pushType: .token)
Task {
    for await data in activity.pushTokenUpdates {
        let myToken = data.map {String(format: "%02x", $0)}.joined()
        // ... required code for entering a live activity
        OneSignal.enterLiveActivity("my_activity_id", withToken: myToken)
    }
}
```
## Functional Testing instructions
Basic case
1. Prepare to make a Live Activity Update

< app_id > : Found on web console
< authorization token for app_id > : Found on web console
< activity_id > : Typed into app at runtime
```
curl --request POST \
     --url https://onesignal.com/api/v1/apps/< app_id >/live_activities/< activity_id >/notifications \
     --header 'Authorization: Basic < authorization token for app_id >' \
     --header 'accept: application/json' \
     --header 'content-type: application/json' \
     --data '
{
     "event": "update",
     "event_updates": {
          "message" : "Hello from OneSignal!"
     },
     "name": "test update"
}
'

```

2. Run the app
3. Scroll to the bottom
3. Type in the chosen activity_id
4. Touch enter live activity
5. Pull down from the upper left
6. See the live activity
7. Run the CURL command above
8. See update!

Case 2: Delayed subscription id
in this case the subscription_id is not available so we need to cache the enter live activity request

in this case we remove
```
+ (void) initialize {
 subscriptionId = OneSignalUserManagerImpl.sharedInstance.pushSubscriptionId;
           
}
```
and add
```
dispatch_async(dispatch_get_main_queue(), ^{
        [NSTimer scheduledTimerWithTimeInterval:30 repeats:NO block:^(NSTimer *timer) {
            subscriptionId = OneSignalUserManagerImpl.sharedInstance.pushSubscriptionId;
            [OneSignalLiveActivityController executePendingLiveActivityUpdates];
        }];
       });
```
at the engining of enterLiveActivity inside the `OneSignalLiveActivityController

 1. Start a live activity
 2. Quickly make an update and see that it does not go through
 3. Wait 30 seconds. (add a breakpoint to the timer block if you want to be sure)
 4. Make an update and see that it now goes through.

## Unit testing
Added several tests to test the requests and the public SDK method

## Manual testing
I have tested Live Activity updates and they work properly.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [x] Public API changes

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [ ] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1189)

### CODE REVIEW

**Review by emawby — CHANGES_REQUESTED** [review]
> Reviewed 26 of 26 files at r1.
> *[Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-iOS-SDK/1189)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @fhboswell, @jennantilla, @jkasten2, and @nan-li)

___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.m` line 398 at r1](https://reviewable.io/reviews/OneSignal/OneSignal-iOS-SDK/1189#-NLXeztC55T-5aet75Hj:-NLXeztC55T-5aet75Hk:b-981qz) ([raw file](https://github.com/OneSignal/OneSignal-iOS-SDK/blob/2da13792fed6ede6798ba9c15910ef250841255e/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L398)):*
> ```objc
> #pragma mark: LIVE ACTIVITIES
> 
> + (void)enterLiveActivity:(NSString * _Nonnull)activityId withToken:(NSString * _Nonnull)token {
> ```

For the major release branch can we move the logic of enter and exit live activities to their own class (OneSignalLiveActivityController?)? I want to keep OneSignal.m as just an entrypoint for public apis. I assume the pending update tracking could also move to this new file if we made it a subscription observer

File: iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+};
+@end
+
+@implementation OneSignalLiveActivityController
+
+static NSMutableArray* pendingLiveActivityUpdates;
+static NSString* subscriptionId;
+
++ (void) initialize {
+    subscriptionId = OneSignalUserManagerImpl.sharedInstance.pushSubscriptionId;
```
Comment (nan-li) [inline iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m:53]: You'll need to add `OneSignalLiveActivityController` as an observer of the push subscription with `[OneSignalUserManagerImpl.sharedInstance addObserver:self]` or some singleton of `OneSignalLiveActivityController` or else `onOSPushSubscriptionChangedWithStateChanges` will never fire for this class.

**Review by nan-li — CHANGES_REQUESTED** [review]

**Comment by nan-li** [comment]
The hacky way I tested pending requests was to comment out the setter in `initialize` and make sure I was optedIn to push with `[OneSignal.User.pushSubscription optIn]` in the end of didFinishLaunchingWithOptions, then I set one of the dev app's buttons to trigger `[OneSignal.User.pushSubscription optOut]`. This will fire that the push subscription changed (even though the subscription ID did not change). But that's enough to get the thing going.

File: iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m (hunk 1 above)
Comment (fhboswell) [reply iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m:61]: Done

File: iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    return self;
+};
+@end
+
+@implementation OneSignalLiveActivityController
+
+static NSMutableArray* pendingLiveActivityUpdates;
+static NSString* subscriptionId;
+
++ (void) initialize {
```
[the remark this answers is not in the document]
Comment (fhboswell) [reply iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m]: Fixed

File: iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    }
+    [pendingLiveActivityUpdates addObject:pendingLiveActivityUpdate];
+}
+
++ (void)executePendingLiveActivityUpdates {
+    subscriptionId =  OneSignalUserManagerImpl.sharedInstance.pushSubscriptionId;
+    if(pendingLiveActivityUpdates.count <= 0) {
+        return;
+    }
+
```
[the remark this answers is not in the document]
Comment (fhboswell) [reply iOS_SDK/OneSignalSDK/Source/OneSignalLiveActivityController.m:158]: Done

**Comment by fhboswell** [comment]
Thank you for the hacky test suggestion. After my changes testing like that fires `onOSPushSubscriptionChangedWithStateChanges`.

**Review by nan-li — APPROVED** [review]
LGTM, maybe we need the license copyright.

**Comment by fhboswell** [comment]
Good call, added legal header on all files.

**Review by emawby — APPROVED** [review]

**Pull request merged**